mbed_host_tests/host_tests_runner/mbed_base.py

```diff
@@ -92,10 +113,69 @@ def copy_image(self, image_path=None, disk=None, copy_method=None, port=None):
             port = self.port
+        if not retry_copy:
+            retry_copy = self.retry_copy
+
+        flash_done = False
+        target_id = self.target_id
+
+        for count in range(0, retry_copy):
+            if flash_done:
+                break
+            initial_remount_count = get_remount_count(disk)
+
+            # Call proper copy method
+            result = self.copy_image_raw(image_path, disk, copy_method, port)
+            sleep(self.program_cycle_s)
+            if not result:
+                continue
+
+            if not target_id:
+                self.logger.prn_wrn("Target ID not found: Skipping flash check and retry")
+                break
+
+            bad_files = Set(['FAIL.TXT'])
+
+                # Re-try at max 5 times with 0.5 sec in delay
+            for i in range(5):
+                # mbed_lstools.create() should be done inside the loop. Otherwise it will loop on same data.
+                mbeds = mbed_lstools.create()
+                mbeds_by_tid = mbeds.list_mbeds_by_targetid()   # key: target_id, value mbedls_dict()
+
+                if target_id in mbeds_by_tid:
+                    if 'mount_point' in mbeds_by_tid[target_id] and mbeds_by_tid[target_id]['mount_point']:
+                        if not initial_remount_count is None:
+                            new_remount_count = get_remount_count(disk)
+                            if not new_remount_count is None and new_remount_count == initial_remount_count:
+                                sleep(0.5)
+                                continue
+
+                        common_items = []
+                        try:
+                            items = Set([x.upper() for x in os.listdir(mbeds_by_tid[target_id]['mount_point'])])
+                            common_items = bad_files.intersection(items)
+                        except OSError as e:
+                            print "Failed to enumerate disk files, retrying"
+                            continue
+
+                        for common_item in common_items:
+                            full_path = os.path.join(mbeds_by_tid[target_id]['mount_point'], common_item)
+                            self.logger.prn_err("Found %s"% (full_path))
+                            bad_file_contents = "[failed to read bad file]"
+                            try:
+                                with open(full_path, "r") as bad_file:
+                                    bad_file_contents = bad_file.read()
+                            except IOError as error:
+                                self.logger.prn_err("Error opening '%s': %s" % (full_path, error))
+
+                            self.logger.prn_err("Error file contents:\n%s"% (bad_file_contents))
+                        if common_items:
+                            result = False
+                        else:
+                            flash_done = True
+                        break
+                    sleep(0.5)
```

Comment:
This sleep should be at the level of ```if target_id in mbeds_by_tid:```. Because if mbedls does not detect this target then we quickly run out of retries.